includes/payment-methods/class-wc-stripe-upe-payment-gateway.php

```diff
@@ -277,16 +278,208 @@ public function get_upe_available_payment_methods() {
+		$selected_upe_payment_type = ! empty( $_POST['wc_stripe_selected_upe_payment_type'] ) ? wc_clean( wp_unslash( $_POST['wc_stripe_selected_upe_payment_type'] ) ) : ''; // phpcs:ignore WordPress.Security.NonceVerification.Missing
+
+		if ( $payment_intent_id ) {
+			if ( $payment_needed ) {
+				$request = [
+					'amount'   => $converted_amount,
+					'currency' => $currency,
+				];
+
+				if ( '' !== $selected_upe_payment_type ) {
+					// Only update the payment_method_types if we have a reference to the payment type the customer selected.
+					$request['payment_method_types'] = [ $selected_upe_payment_type ];
+				}
+
+				if ( $save_payment_method ) {
+					$request['setup_future_usage'] = 'off_session';
+				}
+
+				WC_Stripe_API::request_with_level3_data(
+					$request,
+					"payment_intents/$payment_intent_id",
+					$this->get_level3_data_from_order( $order ),
+					$order
+				);
+			}
+			//} elseif ( $token ) {
+			//	return $this->process_payment_using_saved_method( $order_id );
+		}
+
+		return [
+			'result'         => 'success',
+			'payment_needed' => $payment_needed,
+			'redirect_url'   => wp_sanitize_redirect(
+				esc_url_raw(
+					add_query_arg(
+						[
+							'order_id'            => $order_id,
+							'wc_payment_method'   => self::ID,
+							'_wpnonce'            => wp_create_nonce( 'wcpay_process_redirect_order_nonce' ),
+							'save_payment_method' => $save_payment_method ? 'yes' : 'no',
+						],
+						$this->get_return_url( $order )
+					)
+				)
+			),
+		];
+	}
+
+	/**
+	 * Processes UPE redirect payments.
+	 *
+	 * @param int    $order_id The order ID being processed.
+	 * @param string $intent_id The Stripe setup/payment intent ID for the order payment.
+	 * @param bool   $save_payment_method Boolean representing whether payment method for order should be saved.
+	 *
+	 * @since x.x.x
+	 * @version x.x.x
+	 */
+	public function process_upe_redirect_payment( $order_id, $intent_id, $save_payment_method ) {
+		$order = wc_get_order( $order_id );
+
+		if ( ! is_object( $order ) ) {
+			return;
+		}
+
+		if ( $order->has_status( [ 'processing', 'completed', 'on-hold' ] ) ) {
+			return;
+		}
+
+		WC_Stripe_Logger::log( "Begin processing UPE redirect payment for order $order_id for the amount of {$order->get_total()}" );
+
+		try {
+			// Get user/customer for order.
+			$customer_id = $this->get_stripe_customer_id( $order );
+
+			$payment_needed = 0 < $order->get_total();
+
+			// Get payment intent to confirm status.
+			if ( $payment_needed ) {
+				$intent                 = WC_Stripe_API::retrieve( 'payment_intents/' . $intent_id . '?expand[]=payment_method' );
+				$charge                 = 0 < $intent->charges->total_count ? end( $intent->charges->data ) : null;
+				$payment_method_details = (array) $charge->payment_method_details;
+				$payment_method_type    = $payment_method_details['type'];
```

Comment:
I remember encountering an issue where if there is an error on the intent at this point, then this `payment_method_details` object is empty and referring to the `type` property on null threw an unhappy error and broke things. Consequently, I think we need to check whether `$payment_method_details` is null before referencing its `type` to avoid this error. The same goes for the `$payment_method_options` variable below.

Reply:
Good point, I refactored the code to be a bit more defensive (and also to remove the `Variable 'X' is probably undefined` warnings)

Reply:
Cool, I like the way you handled this one. Looking shnazzy.